core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -341,6 +341,24 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
     }
   }
 
+  /**
+   * Reset the state of CoarseGrainedSchedulerBackend to the initial state. Currently it will only
+   * be called in the yarn-client mode where AM is registered again, also when dynamic
+   * allocation is enabled.
+   * */
+  protected def reset(): Unit = synchronized {
+    if (Utils.isDynamicAllocationEnabled(conf)) {
+      numPendingExecutors = 0
+      executorsPendingToRemove.clear()
+
+      // Remove all the lingering executors that should be removed but not yet. The reason might be
+      // because (1) disconnected event is not yet received; (2) executors die silently.
+      executorDataMap.foreach { case (eid, _) =>
+        driverEndpoint.askWithRetry[Boolean](RemoveExecutor(eid, SlaveLost()))
```

Comment:
Also, I see you're just copying this from the other places, but `RemoveExecutor` doesn't have a reply, so `driverEndpoint.send` should be used instead of `driverEndpoint.askWithRetry`. Probably good to fix the other places in this file where the same is being done.